### DISCUSSION TITLE
More examples.

### DISCUSSION THREAD

**Discussion by excitedbox** [alpinejs/alpine#1108] [4. General]
I think for Alpine to get any kind of adoption it needs many more and CLEAR examples. It is very different for designers when compared to adding a simple on click event that calls a function and letting the devs handle the rest. This moves more of the event functions into the template so very eli5 examples are needed to bring people up to speed. The current examples show one thing while the description shows another which gets confusing.

**Comment by KevinBatdorf** [comment]
I know this doesn't necessarily answer your request, but you can also find a lot of demos on CodePen. 

You can start with pens tagged with Alpinejs https://codepen.io/tag/alpinejs

And check out my collection as well (shameless plug I suppose!): https://codepen.io/collection/nRxrPk

**Reply by excitedbox** [reply]
Yes but examples show you some things and don't give you concrete explanations of the broader system (like looking at a picture from 30ft away you will only see a fraction). Things such as options/arguments, use cases, Syntax, conditionals (if, else, and , or, etc is a common question I have seen) and only somewhat understand myself. Even a simple description of when to use { [ ] } ? : ; etc is missing which would take under 10 minutes to add for someone who is familiar with it. But is non standard syntax for people who have never worked with Vue and will cause many errors and frustration along with wasted time.

The documentation right now is code fragments that are almost impossible to understand without a working example and the 2nd half of the code.  The language used to describe things is also not clear. It is like a race car driver explaining race cars to another race car driver. They understand but nobody else does.

If I invent a language and give you str(word,z) and say "that is used to work with strings", that doesn't tell you what is happening, how it is used, or when to use it.

calling something a descriptor should be followed up with a description of what a descriptor does, how it is used, another word for a descriptor. WHO, WHAT, WHEN, WHERE, WHY, HOW is something we all learned in 2nd grade. The current docs are the What but the rest is missing. 

Most of the examples seem very amateurish and cobbled together. You often see very bad and inefficient ways of solving something instead of using efficient x-for statements etc. Usually the examples use the most basic method because nobody understands the more complex. Very few examples actually demonstrate the true power of Alpine and therefore the demonstrated use cases are very limited. The rest of the more advanced examples all use Alpine to implement JS functions which is really missing the whole point of this.

I would be happy to write up examples but first I need to know how to do things myself. If anybody can get me clear descriptions of how things work, what the options etc are, an like I said CLEARLY explained then I can expand on it with demos and examples.

I know this might seem like harsh criticism but that is not my intention. I am just trying to demonstrate how important it is for good documentation to go with a project. Number one reason people ignore a project is bad or missing documentation. While the second is steep learning curve. The current documentation multiplies the difficulty by 10.  I would hate to see this project fail because I really see potential for it. I actually tossed out a full weeks work and started implementing this instead of UIkit when I found it.

**Reply by KevinBatdorf** [reply]
I'm just an enthusiast and not directly affiliated with the project, but I think the comments are fair. There's room for more of a beginner's guide for Alpine especially if it's to position itself as a jQuery replacer.

One thing I'll point out though is that Alpine is more of a "get out of the way" framework, and assumes JavaScript knowledge. Everything you are probably missing in the docs is likely just regular JavaScript stuff you would learn anywhere. 

Do you think a "from jQuery to Alpine" guide would be something useful?

**Reply by excitedbox** [reply]
I don't think it is basic JS (maybe some things but my points were not just meant to apply to me). Alpine does things more like HTML I would say. One of the problems is that it mixes HTML and JS conventions without clearly explaining when which is used. Also some things apply to one item and not the other so there is no global rules to go by. 

I don't think a guide should assume previous knowledge with Any one system. You would again be writing a guide for a narrow audience while alienating everyone else. What about people who come from Perl/C/Python/Lua etc. 

The other thing is that no project should cater to experts only if it wants to be widely accepted. Nobody has time to learn multiple technologies just to be able to start experimenting with Alpine. Right now you would first need a very solid understanding of JS, JQuery AND Vue to be able to understand the conventions of Alpine. 

Guides should explain how a system works not how it is different from another system. You don't teach someone how to rebuild a Diesel motor by telling them how they are different from a lawnmower engine. I think drawing comparisons is useful when they are used to expand or clarify information. **_They should never rely on you having that other knowledge though._** <-- **This** I see way too often and it leads to the hell that every beginner has to go through when setting up Dev environments or Node. Every guide assumes you are an expert so they all leave out that 1 crucial piece of information that ends up costing you days of debugging hell.

Even a mechanic wouldn't try to rebuild a motor he has never worked on without some kind of documentation.
An example is using :class="" to add a style is not mentioned in the docs so people use convoluted x-show or x-data switching instead of using css to hide or show a components and text. They don't have the information to do even simple things so they loose the efficiency that using Alpine and Tailwind together offers. And those guides right now represent 95% of the teaching materials for the entire Alpine JS ecosystem.

I actually read an article that said if you wan't to do anything more than toggle some basic DOM stuff you should stay away from AlpineJs. In my opinion that is just plain wrong because the fact that Alpine is so simple makes it very powerful. We just need to know how it works to be able to use that power. 


Like I said if someone gets me a table of clear information about how Alpine works I can write up guides for it with examples. That is something someone with experience needs to do though because something that takes them minutes would take me many hours since I would have to test and figure things out myself first. 

Common questions I see are:
- x-if : how to do conditionals especially OR and if else
- Nesting x-for ie when you have an array like in a sub menu where some items might return null.
- x-model could use a lot more info because that will most likely be the most important for inserting components.
- adding or removing css classes ie. sub string manipulation.
- General Syntax and compatibility such as when what can or can not be used.
- How and when to use JS instead of AlpineJS
- Best practices *These 2 are very important for getting people to write efficient and SAFE code*

**Reply by KevinBatdorf** [reply]
Well, I think you have fair points. One thing I'd argue though is that Alpine is designed to make as much sense as possible to beginners by taking the best parts of jQuery, Vue, etc and not assuming knowledge there, all while keeping the codebase super small by only adding in what's necessary. You do need JS knowledge though. Maybe TailwindCSS is a good comparison. Tailwind is pretty intuitive after you start using it, but it assumes you have some knowledge in CSS too. But I think your point might be that Tailwind has some pretty extensive (and amazing) documentation though.

I can try to answer your question:

> x-if : how to do conditionals especially OR and if else

You can't. You would just use a second `x-if` and reverse the logic. This was intentional to keep the code smaller.

> Nesting x-for ie when you have an array like in a sub menu where some items might return null.

This probably works how you'd expect already, but for null values you should filter those out with regular JS. ie. `<div x-for="items in items.filter(Boolean)">`

> x-model could use a lot more info because that will most likely be the most important for inserting components.

Can't argue with that.

> adding or removing css classes ie. sub string manipulation.

yeah, probably needs more examples there too. Here's a couple I shared recently on Discord:
```html
<span :class="step == 3 ? 'text-indigo-600' : 'text-gray-500 group-hover:text-gray-900'">03</span>
```
and
```html
<span :class="{
  'text-indigo-600': step == 2,
  'text-gray-500 group-hover:text-gray-900': step != 2,
  'another-class': anotherCondition,
}">02</span>
```

> General Syntax and compatibility such as when what can or can not be used.

I don't necessarily get what you mean here, but I think maybe you are thinking Alpine does more than it actually does. One example is the event system. For the most part, Alpine only gives you two things, a way to fire events `@`, and a way to listen to them `x-on`. With those two, the user can fit in any browser events that are supported. Alpine has no knowledge of these events and just gets out of the way. `@click=""` isn't Alpine specifically. Instead, it's just passing the word "click" to Alpine which happens to be the name of an event the browser supports. 

> How and when to use JS instead of AlpineJS

I think generally, anything inside a directive (`x-on`, `x-if`, `x-show`) you're using pure JS unless you see something starting with a `$` (`$event`, `$dispatch()`, etc), which Alpine refers to as "magic properties" (or "magic helpers" you might see). I can give you some examples if you have questions about this more.

> Best practices These 2 are very important for getting people to write efficient and SAFE code

Yeah, I agree. There is generally an "Alpine way" of doing things that typically is a declarative/functional approach. I often see this as confusing to users so there's definitely room for improvement in communication here too.

---

If you have more questions I can try to answer them the best I can. I'll say it again though, _I'm just a user and not the author, so my comments are just my thoughts and opinions as much as yours are!_

You also might want to hold off a bit if investing time into making something though until V3 is released. Caleb mentioned some new documentation being released at that time, so maybe a lot of the stuff you find is missing will be there!

**Reply by excitedbox** [reply]
In you step example is that turning the :css into ternary? or does it need a x-if before that?

For my example of filtering empty nested items wouldn't it be easier to do x-if="item != null" :x-for="item in items" .... or something like that to eliminate the need for writing JS.

**Comment by excitedbox** [comment]
Thanks for the reply that gives me a good starting point. Yea tailwind assumes you know css but their documentation gives plenty of details that a CSS beginner can get up to speed. While Alpine's documentation does require a pretty substantial amount of experience or you are sent on a wild goose chase. Especially with coding that is a bad place to be because nailing down bugs is so difficult. 

I will try to pull this together and write something up.

What I meant by syntax is grammar rules. A lot of people are getting confused when to follow JS conventions and when to use CSS/HTML. Which is why I think it needs to be clarified. I am trying to look at this from the perspective of someone like my boss who knows HTML, CSS and just enough JS to include a lightbox in a page. I think for those people the simplicity that alpine brings will be very useful but the documentation needs to show them how simple things can be.

To me syntax is things like in an x-data when you are listing items you would use:  item: { } or item: [ ] or item { [ ] } or item [ { } ] depending on the type of data/variable/ array. When to put item: 'value' or item: value and the ternary ? : operator usage (when / where). Where you can add :class (or any other unlisted possibilities). 
I can pull those from the JS guide at w3schools.

Vue has this style guide which I think is important to have to make sure people write safe and efficient code. Since I just started using Alpine I don't think I am qualified to write this yet since I am still getting used to the differences. I have been working with php and mysql 99% of the time for 17 years so client side is not my thing.
https://vuejs.org/v2/style-guide/

I will start with these sections for now and see how far I get:
- syntax/ operators, ie. formatting rules
- explaining how to use/ format x-data items (maybe I will add a simple ajax example)
- Manipulating css
- - adding / removing classes (I guess that could go with the x-text inner/html stuff)
- - sub string manipulation
- Expand on the x-if text that is in the current docs
- Style Guide / Best Practices (I will try to get this started but others will need to expand it)
